tortoise/tortoise.go

```diff
@@ -904,40 +831,112 @@ func (t *turtle) layerCutoff() types.LayerID {
 	return t.last.Sub(t.Hdist)
 }
 
-func isUndecided(config Config, decided map[types.LayerID]types.BlockID, lid, last types.LayerID) bool {
-	if _, exists := decided[lid]; exists {
-		return false
+func (t *turtle) decodeExceptions(base, ballot *ballotInfo, exceptions types.Votes) {
+	from := base.layer
+	diff := map[types.LayerID]map[types.BlockID]sign{}
+	for vote, bids := range map[sign][]types.BlockID{
+		support: exceptions.Support,
+		against: exceptions.Against,
+	} {
+		for _, bid := range bids {
+			block, exist := t.blockRefs[bid]
+			if !exist {
+				ballot.conditions.votesBeforeBase = true
+				continue
+			}
+			if block.layer.Before(from) {
+				ballot.conditions.votesBeforeBase = true
+				from = block.layer
+			}
+			layerdiff, exist := diff[block.layer]
+			if !exist {
+				layerdiff = map[types.BlockID]sign{}
+				diff[block.layer] = layerdiff
+			}
+			layerdiff[block.id] = vote
+		}
 	}
-	genesis := types.GetEffectiveGenesis()
-	limit := genesis
-	if last.After(genesis.Add(config.Zdist)) {
-		limit = last.Sub(config.Zdist)
+	for _, lid := range exceptions.Abstain {
+		if lid.Before(from) {
```

Comment:
from could be modified from line 849. why not directly compare with base.layer?

Reply:
same as above